File: Tarefa3/scene_graph/python/luxor/animation.py

class Animation:
  def __init__ (self, moves):
    self.curr = 0
    self.moves = moves.copy()

  # return true when its done
  def Advance (self, dt, reverse=False):
    if reverse:
      idx = (len(self.moves)-1) - self.curr 
    else:
      idx = self.curr
    if self.moves[idx].Advance(dt,reverse):
      self.curr += 1
      if self.curr == len(self.moves):
        self.curr = 0
        return True
    return False

File: projeto3/src/luxor/animation.cpp
#include "animation.h"

  std::vector<MovementPtr> m_moves;
Animation::Animation (std::initializer_list<MovementPtr> moves)
: m_curr(0), m_moves(moves)
{
}
AnimationPtr Animation::Make (std::initializer_list<MovementPtr> moves)
{
  return AnimationPtr(new Animation(moves));
}
Animation::~Animation ()
{
}

bool Animation::Advance (float dt, bool reverse)  // return true when it's done
{
  // use size_t to avoid signed/unsigned comparison warnings
  size_t idx = reverse ? (m_moves.size()-1)-m_curr : m_curr;
  if (m_moves[idx]->Advance(dt,reverse)) {
    ++m_curr;
    if (m_curr >= m_moves.size()) { // finished cycle
      m_curr = 0;
      return true;
    }
  }
  return false;
}
